### DISCUSSION TITLE
Fixing code quality issues with static analysis

### DISCUSSION THREAD

**Discussion by de-sh** [Project-MONAI/MONAI#1404] [Ideas]
Hi, I was going through the code and decided to run it through [DeepSource](https://deepsource.io), and found some interesting results. You might want to have a look for yourself as it detected [292 performance issues](https://deepsource.io/gh/de-sh/MONAI/issues/?category=performance), [641 bug risks](https://deepsource.io/gh/de-sh/MONAI/issues/?category=bug-risk) and [116 potential security vulnerabilities](https://deepsource.io/gh/de-sh/MONAI/issues/?category=security). Consider taking a look at the [report generated by DeepSource](https://deepsource.io/gh/de-sh/MONAI/).

I was considering opening a PR to solve these issues, shall I?

**Comment by wyli** [comment]
is it possible to check every pull request and report automatically?
if you want to resolve those please make separate PRs for different types of issues (so that they are easy for review)

**Reply by de-sh** [reply]
Hey @wyli, yes!
DeepSource does performs continuous static analysis and integrates well into the PR workflow on GitHub/GitLab/BitBucket.
Ok, I shall open separate PRs to solve the issues, by type.

**Reply by de-sh** [reply]
I have opened a PR(#1423) to fix some of the performance issues highlighted by DeepSource. While you consider merging the PR, I strongly recommend that you enable DeepSource on the repo, as it is free for OpenSource projects. If you are interested, please follow these steps:

- Sign up on [DeepSource](https://deepsource.io) with your GitHub account and grant access to this repo.
- Activate analysis on this repo.
- You can also look at the docs for more details. Do let me know if I can be of any help!

**Reply by wyli** [reply]
I feel we could enable this tool, mainly to make coding style more consistent, improve the readability. what do you think? @Nic-Ma @ericspod @rijobro 

(it's free for OSS)

**Reply by rijobro** [reply]
Anything that improves code quality is good for me!

**Reply by wyli** [reply]
thanks I've installed the app https://deepsource.io/gh/Project-MONAI/MONAI/ and created a demo autofix  https://github.com/Project-MONAI/MONAI/pull/1427 pls let me know for any issues

**Reply by ericspod** [reply]
I think this can be useful for spotting a number of things. There's a few things it found that are genuine code quality issues, some flagging arguments or syntax that is not needed but might be helpful for readability. One thing it doesn't understand properly is the `TestCase` derived classes and why test cases need to be methods even if `self` isn't used, but I guess we can set some config to filter out the tests directory.

**Reply by ericspod** [reply]
There is also a few debatable choices for what's considered an issue. One thing that's mentioned is " Ideally, assert statement should be used only in tests." This gets flagged whenever an assert is used outside a test, but I would say that an assert to check for sanity mid-algorithm is one of the intended use cases. You shouldn't rely on the check to validate input or otherwise assume the assert is always there, but it has its place outside tests.

**Reply by tvercaut** [reply]
In previous projects I worked on, we had an `assert_or_raise` function that would `assert` in debug mode but `raise` in production mode. The rationale was to avoid dramatic exits of the software or undefined/untested behavior (as asserts are typically simply disabled by [`-O` or explicitly setting `__debug__` to `false`](https://docs.python.org/3/library/constants.html#__debug__) in python or `#define NDEBUG` in C++) in case the asserted condition was not met during real uses. A more graceful error message could be shown to the user in many cases.

**Reply by srijan-deepsource** [reply]
> but I guess we can set some config to filter out the tests directory

@ericspod That is right. DeepSource silences many issues that don't make sense in test files. We just [need a glob](https://deepsource.io/docs/config/deepsource-toml.html#test-patterns) to recognize the files. 

Also, about `assert` statements: the idea behind the issue is the same that @tvercaut mentioned.

**Reply by rijobro** [reply]
Hi @tvercaut, great idea, but I guess we wouldn't want to change all `assert` to `assert_or_raise`. In previous code I worked on (C++), we had an assert when accessing array data to check that each of the indices were in bounds. This was considerably slower, and was only really necessary when debugging.

**Reply by tvercaut** [reply]
@rijobro Indeed, in time critical code, we also were not converting asserts into exceptions but just skipped these when in production mode.

**Reply by ericspod** [reply]
@srijan-deepsource OK makes sense, that looks like an easy change. My feeling with asserts though is whenever you would want an exception raise in production you would want that in testing as well. Assert is for those cases where you would want a check in test but not in production because your code is considered correct and the check isn't needed. This is a minor point but in general the feedback from the tools looks very helpful.
